main/src/library/Map.flix

```diff
@@ -1,7 +1,335 @@
+    def union[k, v](m1: Map[k, v], m2: Map[k, v]): Map[k, v] = unionWithKey((k, v1, v2) -> v1, m1, m2)
+
+    ///
+    /// Returns the union of `m1` and `m2` where key collisions are resolved with the merge function `f`.
+    ///
+    def unionWith[k, v](f: (v, v) -> v, m1: Map[k, v], m2: Map[k, v]): Map[k, v] = unionWithKey((k, v1, v2) -> f(v1, v2), m1, m2)
+
+    ///
+    /// Returns the union of `m1` and `m2` where key collisions are resolved with the merge function `f`, taking both the key and values.
+    ///
+    def unionWithKey[k, v](f: (k, v, v) -> v, m1: Map[k, v], m2: Map[k, v]): Map[k, v] =
+        foldRightWithKey((k, v, m) -> insertWithKey((key, v1, v2) -> f(key, v1, v2), k, v, m) , m2, m1)
+
+    ///
+    /// Returns the left-biased intersection of `m1` and `m2`.
+    /// That is, key collisions are resolved by taking the mapping from `m1`.
+    ///
+    def intersection[k, v](m1: Map[k, v], m2: Map[k, v]): Map[k, v] = intersectionWithKey((k, v1, v2) -> v1, m1, m2)
+
+    ///
+    /// Returns the intersection of `m1` and `m2` where key collisions are resolved with the merge function `f`.
+    ///
+    def intersectionWith[k, v](f: (v, v) -> v, m1: Map[k, v], m2: Map[k, v]): Map[k, v] = intersectionWithKey((k, v1, v2) -> f(v1, v2), m1, m2)
+
+    ///
+    /// Returns the intersection of `m1` and `m2` where key collisions are resolved with the merge function `f`, taking both the key and values.
+    ///
+    def intersectionWithKey[k, v](f: (k, v, v) -> v, m1: Map[k, v], m2: Map[k, v]): Map[k, v] =
+        let intersect = filterWithKey((k, v) -> memberOf(k, m1), m2);
+        foldRightWithKey((k, v, m) -> adjustWithKey((key, v1) -> f(key, v, v1), k, m), intersect, m1)
+
+    ///
+    /// Returns the difference of `m1` and `m2`, i.e. `m1 - m2`.
+    /// That is, returns the map `m1` with the keys removed that are in `m2`.
+    ///
+    def difference[k, v](m1: Map[k, v], m2: Map[k, v]): Map[k, v] = differenceWithKey((k, v1, v2) -> None, m1, m2)
+
+    ///
+    /// Returns the difference of `m1` and `m2`, i.e. `m1 - m2`.
+    /// When a key `k` is in both `m1` and `m2`, the associated values are passed to the merge function `f`.
+    /// If `f` returns `None` the mapping with `k` is thrown away (proper set difference).
+    /// If `f` returns `Some(v)` the mapping `k -> v` is included in the result.
+    ///
+    def differenceWith[k, v](f: (v, v) -> Option[v], m1: Map[k, v], m2: Map[k, v]): Map[k, v] = differenceWithKey((k, v1, v2) -> f(v1, v2), m1, m2)
+
+    ///
+    /// Returns the difference of `m1` and `m2`, i.e. `m1 - m2`.
+    /// When a key `k` is in both `m1` and `m2`, `k` and the associated values are passed to the merge function `f`.
+    /// If `f` returns `None` the mapping with `k` is thrown away (proper set difference).
+    /// If `f` returns `Some(v)` the mapping `k -> v` is included in the result.
+    ///
+    def differenceWithKey[k, v](f: (k, v, v) -> Option[v], m1: Map[k, v], m2: Map[k, v]): Map[k, v] =
```

Comment:
I haven't been able to figure out a cleaner way of doing this, any thoughts?  At the very least, I think it is clear the way it is.